container/lru/map.go

```diff
@@ -0,0 +1,524 @@
+	m.mtx.Lock()
+
+	elem, exists := m.items[key]
+	if !exists || m.isElemExpired(elem, m.nowFn()) {
+		m.misses++
+		return *new(V), false
+	}
+
+	m.hits++
+	m.moveElemToFront(elem)
+	return elem.value, true
+}
+
+// Exists returns whether or not the passed item is a member.  The priority and
+// expiration time for the item are not modified.  It does not affect the hit
+// ratio.
+//
+// This function is safe for concurrent access.
+func (m *Map[K, V]) Exists(key K) bool {
+	m.mtx.Lock()
+	elem, exists := m.items[key]
+	exists = exists && !m.isElemExpired(elem, m.nowFn())
+	m.mtx.Unlock()
+	return exists
+}
+
+// Peek returns the associated value of the passed key if it is a member without
+// modifying any priority or the hit ratio.  The boolean return value indicates
+// whether or not the item lookup is successful.
+//
+// This function is safe for concurrent access.
+func (m *Map[K, V]) Peek(key K) (V, bool) {
+	defer m.mtx.Unlock()
+	m.mtx.Lock()
+
+	elem, exists := m.items[key]
+	if !exists || m.isElemExpired(elem, m.nowFn()) {
+		return *new(V), false
+	}
+
+	return elem.value, true
+}
+
+// Len returns the number of items in the map.
+//
+// This function is safe for concurrent access.
+func (m *Map[K, V]) Len() uint32 {
+	m.mtx.Lock()
+	numItems := uint32(len(m.items))
+	m.mtx.Unlock()
+	return numItems
+}
+
+// Delete deletes the item associated with the passed key if it exists.
+//
+// This function is safe for concurrent access.
+func (m *Map[K, V]) Delete(key K) {
+	m.mtx.Lock()
+	if elem, exists := m.items[key]; exists {
+		m.deleteElem(key, elem)
+	}
+	m.mtx.Unlock()
+}
+
+// EvictExpiredNow immediately removes all items that are marked expired without
+// waiting for the next expiration scan.  It returns the number of items that
+// were removed.
+//
+// It is expected that most callers will typically use the automatic lazy
+// expiration behavior and thus have no need to call this method.  However, it
+// is provided in case more deterministic expiration is required.
+func (m *Map[K, V]) EvictExpiredNow() (numEvicted uint32) {
+	defer m.mtx.Unlock()
+	m.mtx.Lock()
+
+	if !m.hasTimeouts {
+		return 0
+	}
+
+	numEvicted = m.removeExpired(m.nowFn())
+	return numEvicted
+}
+
+// Clear removes all items and resets the hit ratio.
+//
+// This function is safe for concurrent access.
+func (m *Map[K, V]) Clear() {
+	m.mtx.Lock()
+	for k, elem := range m.items {
+		m.deleteElem(k, elem)
+	}
+	m.root.prev = &m.root
+	m.root.next = &m.root
+	m.hasTimeouts = m.ttl > 0
+	m.nextExpireScan = m.nowFn().Add(expireScanInterval)
+	m.hits = 0
+	m.misses = 0
+	m.mtx.Unlock()
+}
+
+// Keys returns a slice of unexpired keys ordered from least recently used to
+// most recently used.  The priority and expiration times for the items are not
+// modified.
+//
+// This function is safe for concurrent access.
+func (m *Map[K, V]) Keys() []K {
+	defer m.mtx.Unlock()
+	m.mtx.Lock()
+
+	numItems := len(m.items)
+	if numItems == 0 {
+		return nil
```

Comment:
Returning an empty slice instead of nil could help to prevent panics in the calling code. Especially given the comment which says the func will return a slice and doesnt mention the chance of returning nil.

Same applies to `Values`.

Reply:
`nil` is a value of a slice.

What would the caller's usage look like here, where they improperly index into a nil slice but not an empty non-nil slice? Are they checking s != nil instead of len(s) != 0?

Reply:
Checking s != nil seems like it would be *more* prone to panics.

```
s := []int{} // returned from func
// later...
if s != nil {
    _ = s[0] // boom
}
```

Reply:
With your comments and also looking at a bit at [the go wiki](https://go.dev/wiki/CodeReviewComments#declaring-empty-slices), I think you are right and the code is fine as-is.

Reply:
Already hinted at above, but in addition to pretty much always returning `nil` for empty slices per the linked wiki entry, as a general rule, you should also pretty much always use `if len(s) > 0` instead of checking for `nil` slices explicitly because `len(s)` is 0 both for empty slices and nil slices and thus you cover both cases with a single check which is almost always safer.  After all, if you're checking the length, it's a good bet you're about to access it and so you really need to make sure it's both non-nil _and_ it has enough items for the index you intend to access.